From 7ebba1a9631aeb72cf24d771533f5938eda4b0e8 Mon Sep 17 00:00:00 2001
From: Rudi Heitbaum <rudi@heitbaum.com>
Date: Sat, 8 Aug 2026 12:00:00 +0000
Subject: [PATCH] PY_LGPIO: replace removed SWIG Python 2 compatibility macros

swig used to write a "Compatibility macros for Python 3" block into
every generated wrapper, aliasing the Python 2 spellings onto their
Python 3 names. swig 4.5.0 removed that block, so the typemaps in
lgpio.i no longer compile:

  lgpio_wrap.c:3759:12: error: implicit declaration of function
    'PyInt_FromLong'; did you mean 'PyLong_FromLong'?
  lgpio_wrap.c:3760:12: error: implicit declaration of function
    'PyString_FromString'; did you mean 'PyLong_FromString'?
  lgpio_wrap.c:3760:10: error: assignment to 'PyObject *' from 'int'
    makes pointer from integer without a cast

Call the Python 3 functions directly:

- PyInt_Check          -> PyLong_Check
- PyInt_AsLong         -> PyLong_AsLong
- PyInt_FromLong       -> PyLong_FromLong
- PyString_FromString  -> PyUnicode_FromString

The replacements are exactly what swig aliased them to, so there is no
behaviour change, and they are valid back to Python 2.7.

Link: https://github.com/swig/swig/commit/79f7a2b7cb7d

Signed-off-by: Rudi Heitbaum <rudi@heitbaum.com>
---
 PY_LGPIO/lgpio.i | 42 +++++++++++++++++++++---------------------
 1 file changed, 21 insertions(+), 21 deletions(-)

diff --git a/PY_LGPIO/lgpio.i b/PY_LGPIO/lgpio.i
index 5f1bfeb..bead4e3 100644
--- a/PY_LGPIO/lgpio.i
+++ b/PY_LGPIO/lgpio.i
@@ -296,12 +296,12 @@ error_text                Get the error text for an error code
 // lgSpiRead
 %typemap(in) (char *rxBuf, int count)
 {
-   if (!PyInt_Check($input))
+   if (!PyLong_Check($input))
    {
       PyErr_SetString(PyExc_ValueError, "Expecting an integer");
       SWIG_fail;
    }
-   $2 = PyInt_AsLong($input);
+   $2 = PyLong_AsLong($input);
    if ($2 < 0)
    {
       PyErr_SetString(PyExc_ValueError, "Positive integer expected");
@@ -360,7 +360,7 @@ error_text                Get the error text for an error code
    Py_XDECREF($result);   /* Blow away any previous result */
 
    $result = PyList_New(2);
-   o1 = PyInt_FromLong(result);
+   o1 = PyLong_FromLong(result);
    PyList_SetItem($result, 0, o1);
 
    if (result > 0) o2 = PyByteArray_FromStringAndSize($2, result);
@@ -408,7 +408,7 @@ error_text                Get the error text for an error code
    PyObject *o1, *o2;
    Py_XDECREF($result);   /* Blow away any previous result */
    $result = PyList_New(2);
-   o1 = PyInt_FromLong(result);
+   o1 = PyLong_FromLong(result);
    PyList_SetItem($result, 0, o1);
    if (result > 0) o2 = PyByteArray_FromStringAndSize($1, result);
    else o2 = PyByteArray_FromStringAndSize("", 0);
@@ -428,7 +428,7 @@ error_text                Get the error text for an error code
    Py_XDECREF($result);   /* Blow away any previous result */
    $result = PyList_New(2);
 
-   o1 = PyInt_FromLong(result);
+   o1 = PyLong_FromLong(result);
    PyList_SetItem($result, 0, o1);
 
    if (result > 0) o2 = PyByteArray_FromStringAndSize($1, result);
@@ -494,18 +494,18 @@ error_text                Get the error text for an error code
    {
       result = 0;
 
-      o2 = PyInt_FromLong(chipInf2.lines);
-      o3 = PyString_FromString(chipInf2.name);
-      o4 = PyString_FromString(chipInf2.label);
+      o2 = PyLong_FromLong(chipInf2.lines);
+      o3 = PyUnicode_FromString(chipInf2.name);
+      o4 = PyUnicode_FromString(chipInf2.label);
    }
    else
    {
-      o2 = PyInt_FromLong(0);
-      o3 = PyString_FromString("");
-      o4 = PyString_FromString("");
+      o2 = PyLong_FromLong(0);
+      o3 = PyUnicode_FromString("");
+      o4 = PyUnicode_FromString("");
    }
 
-   o1 = PyInt_FromLong(result);
+   o1 = PyLong_FromLong(result);
 
    PyList_SetItem($result, 0, o1);
    PyList_SetItem($result, 1, o2);
@@ -530,20 +530,20 @@ error_text                Get the error text for an error code
    {
       result = 0;
 
-      o2 = PyInt_FromLong(lineInf3.offset);
-      o3 = PyInt_FromLong(lineInf3.lFlags);
-      o4 = PyString_FromString(lineInf3.name);
-      o5 = PyString_FromString(lineInf3.user);
+      o2 = PyLong_FromLong(lineInf3.offset);
+      o3 = PyLong_FromLong(lineInf3.lFlags);
+      o4 = PyUnicode_FromString(lineInf3.name);
+      o5 = PyUnicode_FromString(lineInf3.user);
    }
    else
    {
-      o2 = PyInt_FromLong(0);
-      o3 = PyInt_FromLong(0);
-      o4 = PyString_FromString("");
-      o5 = PyString_FromString("");
+      o2 = PyLong_FromLong(0);
+      o3 = PyLong_FromLong(0);
+      o4 = PyUnicode_FromString("");
+      o5 = PyUnicode_FromString("");
    }
 
-   o1 = PyInt_FromLong(result);
+   o1 = PyLong_FromLong(result);
 
    PyList_SetItem($result, 0, o1);
    PyList_SetItem($result, 1, o2);
